models/migrations/v221.go

```diff
@@ -0,0 +1,75 @@
+// Copyright 2022 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package migrations
+
+import (
+	"code.gitea.io/gitea/modules/timeutil"
+
+	"xorm.io/xorm"
+	"xorm.io/xorm/schemas"
+)
+
+func increaseCredentialIDTo1640(x *xorm.Engine) error {
+	// Create webauthnCredential table
+	type webauthnCredential struct {
+		ID              int64 `xorm:"pk autoincr"`
+		Name            string
+		LowerName       string `xorm:"unique(s)"`
+		UserID          int64  `xorm:"INDEX unique(s)"`
+		CredentialID    string `xorm:"INDEX VARCHAR(1640)"` // CredentialID is at most 1023 bytes as per spec released 20 July 2022 -> 1640 base32 encoding
+		PublicKey       []byte
+		AttestationType string
+		AAGUID          []byte
+		SignCount       uint32 `xorm:"BIGINT"`
+		CloneWarning    bool
+		CreatedUnix     timeutil.TimeStamp `xorm:"INDEX created"`
+		UpdatedUnix     timeutil.TimeStamp `xorm:"INDEX updated"`
+	}
+	if err := x.Sync2(&webauthnCredential{}); err != nil {
+		return err
+	}
+
+	switch x.Dialect().URI().DBType {
+	case schemas.MYSQL:
+		_, err := x.Exec("ALTER TABLE webauthn_credential MODIFY COLUMN credential_id VARCHAR(1640)")
+		if err != nil {
+			return err
+		}
+	case schemas.ORACLE:
+		_, err := x.Exec("ALTER TABLE webauthn_credential MODIFY credential_id VARCHAR(1640)")
+		if err != nil {
+			return err
+		}
+	case schemas.MSSQL:
+		// This column has an index on it. I could write all of the code to attempt to change the index OR
+		// I could just use recreate table.
+		sess := x.NewSession()
+		if err := sess.Begin(); err != nil {
+			_ = sess.Close()
```

Comment:
why not use a defer for sess close?

Reply:
TBH I just copied this code from v210.go and dropped the second part of that migration... and I didn't think more about it as I knew that this part of code has been proven to work in real-world use. This migration is going back to 1.17 and we need to release that quickly I therefore wanted to use as proven a set of a code as possible. 

I mean I guess we could use defer sess.Close() here. 

I guess we could also attempt to extract out a function that does the column resizing and do it properly even for MSSQL but we'll definitely need to write a test if we do that.

Thinking on a valid test for this migration would be - prepare the table to have one or two entries with the old size, ensure the migration doesn't die, and then just need to test insertion of a huge credential ID at the end.

Reply:
probably we'd need two tests the case where the column size is the old size and one where the column size is the new size but I don't think that breaks.

Reply:
@zeripath should I push the defer patch?

Reply:
If you want